doc/source/dev/api-dev/nan_policy.rst

```diff
@@ -0,0 +1,168 @@
+A Design Specification for `nan_policy`
+=======================================
+
+Many functions in `scipy.stats` have a parameter called ``nan_policy``
+that determines how the function handles data that contains ``nan``.  In
+this section, we provide SciPy developer guidelines for how ``nan_policy``
+is intended to be used, to ensure that as this parameter is added to new
+functions, we maintain a consistent API.
+
+The basic API
+-------------
+
+The parameter ``nan_policy`` accepts three possible strings: ``'omit'``,
+``'raise'`` and ``'propagate'``.  The meanings are:
+
+* ``nan_policy='omit'``:
+  Ignore occurrences of ``nan`` in the input.  Do not generate a warning
+  if the input contains ``nan``. For example, for the simple case of a
+  function that accepts a single array (and ignoring the possible use of
+  ``axis`` for the moment)::
+
+      func([1.0, 3.0, np.nan, 5.0], nan_policy='omit')
+
+  should behave the same as::
+
+      func([1.0, 3.0, 5.0])
+
+  More generally, ``func(a, nan_policy='omit')`` should behave the same as
+  ``func(a[~np.isnan(a)])``.
+
+  Unit tests for this property should be used to test functions that
+  handle ``nan_policy``.
+
+  For functions that accept two or more arguments but whose values are
+  not related, the same idea applies to each input array.  So::
+
+      func(a, b, nan_policy='omit')
+
+  should behave the same as::
+
+      func(a[~np.isnan(a)], b[~np.isnan(b)])
+
+  For inputs with *related* values, the recommended behavior is to omit
+  all the values for which any of the related values are ``nan``.  For
+  a function with two related array inputs, this means::
+
+      y = func(a, b, nan_policy='omit')
+
+  should behave the same as::
+
+      hasnan = np.isnan(a) | np.isnan(b)  # Union of the isnan masks.
+      y = func(a[~hasnan], b[~hasnan])
+
+  The docstring for such a function should clearly state this behavior.
+
+* ``nan_policy='raise'``:
+  Raise a ``ValueError``.
```

Comment:
(Getting back to this after a long delay...)

I see that as an implementation detail, and not really part of the API specification.  We might rewrite,  rename, or relocate that function in the future, and that should not change the API spec.  And for some functions, we might not even use that function.  Also, I think of this document as useful for both developers and users, and we don't need to subject the users to those implementation details here.

Having said that, it might be useful to add a section (in this doc, or perhaps a separate one) about the current recommended best practices for implementing the API, with the understanding that the implementation details are subject to change.  To minimize the additional review needed for this PR, I think we can defer that to a later PR.

Reply:
>Also, I think of this document as useful for both developers and users, and we don't need to subject the users to those implementation details here.

This makes more sense, I was reading this as guidance for those of us maintaining/implementing functions that support `nan_policy` and not necessarily for users.  

> Having said that, it might be useful to add a section (in this doc, or perhaps a separate one) about the current recommended best practices for implementing the API, with the understanding that the implementation details are subject to change. To minimize the additional review needed for this PR, I think we can defer that to a later PR.

Agreed, for several reasons, let's avoid implementation details here.